### PR TITLE
Allow IP address/host restrictions using `restrictedAddresses` option

### PR DESCRIPTION
**Pull request by kamalaknn**

### CODE REVIEW

File: lib/requester/util.js [hunk 1]
```diff
@@ -14,7 +27,77 @@ var _ = require('lodash'),
...
+     */
+    lookup = function (lookupOptions, hostname, options, callback) {
+        var self = this;
+
+        if (hostname !== LOCALHOST) {
+            return dns.lookup(hostname, options, function (err, address, family) {
+                if (err) { return callback(err); }
+                callback(
+                    self.isAddressRestricted(address, lookupOptions) ?
+                        new Error('runtime: IP/host has been blocked') :
```
Comment (kunagpal) [inline lib/requester/util.js]: Needs more debate...

File: lib/requester/util.js [hunk 2]
```diff
@@ -14,7 +27,77 @@ var _ = require('lodash'),
...
+
+        async.waterfall([
+            function (next) {
+                // Try checking if we can connect to IPv6 localhost ('::1')
+                connect(LOCAL_IPV6, lookupOptions.port, function (err) {
+                    next(null, !err);
+                });
+            },
+            function (result, next) {
+                return next(null, result ? LOCAL_IPV6 : LOCAL_IPV4, result ? 6 : 4);
```
Comment (kunagpal) [inline lib/requester/util.js]: Duplicate `result` check.

File: lib/requester/util.js [hunk 3]
```diff
@@ -14,7 +27,77 @@ var _ = require('lodash'),
...
+                    self.isAddressRestricted(address, lookupOptions) ?
+                        new Error('runtime: IP/host has been blocked') :
+                        err,
+                    address,
+                    family
+                );
+            });
+        }
+
+        async.waterfall([
```
Comment (kunagpal) [inline lib/requester/util.js]: The `waterfall` can be compacted into a regular callback flow.

File: lib/requester/util.js [hunk 4]
```diff
@@ -74,6 +161,18 @@ module.exports = {
         // it is bubbled up to us after the request is made. If left to the underlying core, it's not :/
         self.ensureHeaderExists(options.headers, 'Host', request.url.getRemote());
 
+        // override DNS lookup
+        if (defaultOpts.restrictedAddresses || hostname === LOCALHOST) {
+            isSSL = _.startsWith(request.url.protocol, 'https');
+            portNumber = port || (isSSL ? '443' : '80');
```
Comment (kunagpal) [inline lib/requester/util.js]: `Number(port) || (isSSL ? 443 : 80)`
`const`

File: lib/requester/util.js [hunk 5]
```diff
@@ -14,7 +33,70 @@ var _ = require('lodash'),
...
+        var self = this;
+
+        if (hostname !== LOCALHOST) {
+            return dns.lookup(hostname, options, function (err, address, family) {
+                if (err) { return callback(err); }
+                callback(
+                    self.isAddressRestricted(address, lookupOptions) ?
+                        new Error(ERROR_RESTRICTED_ADDRRESS) :
+                        err,
+                    address,
```
Comment (shamasis) [inline lib/requester/util.js]: don't add the newlines here.

File: lib/requester/util.js [hunk 6]
```diff
@@ -34,7 +116,11 @@ module.exports = {
         var options = {},
             self = this,
             bodyParams,
-            url;
+            url,
+            isSSL,
+            portNumber,
+            port = _.get(request, 'url.port'),
```
Comment (shamasis) [inline lib/requester/util.js]: remove _.get from core lib

File: lib/requester/util.js [hunk 7]
```diff
@@ -304,5 +401,18 @@ module.exports = {
...
+     * Checks if a given host or IP is has been restricted in the options.
+     *
+     * @param {String} host
+     * @param {Object} options
+     * @param {Array<String>} options.restrictedAddresses
+     *
+     * @returns {Boolean}
+     */
+    isAddressRestricted: function (host, options) {
+        return options.restrictedAddresses && options.restrictedAddresses[host];
```
Comment (shamasis) [inline lib/requester/util.js]: do lowercase here

**Pull request merged**